[BoundsSafety][LLDB] Implement instrumentation plugin for -fbounds-safety soft traps

delcypher opened the pull request:
This patch implements an instrumentation plugin for the
`-fbounds-safety` soft trap mode first implemented in
https://github.com/swiftlang/llvm-project/pull/11645 (rdar://158088757).

The current implementation of -fbounds-safety traps works by emitting
calls to runtime functions intended to log the occurrence of a soft trap.
While the user could just set a breakpoint of these functions the
instrumentation plugin sets it automatically and provides several
additional features:

When debug info is available:

* It adjusts the stop reason to be the reason for trapping. This is
  extracted from the artificial frame in the debug info (similar to
  -fbounds-safety hard traps).
* It adjusts the selected frame to be the frame where the soft trap
  occurred.

When debug info is not available:

* For the `call-with-str` soft trap mode the soft trap reason is
  read from the first argument register.
* For the `call-minimal` soft trap mode the stop reason is adjusted
  to note its a bounds check failure but does not give further
  information because none is available.
* In this situation the selected frame is not adjusted because in
  this mode the user will be looking at assembly and adjusting the
  frame makes things confusing.

This patch includes shell and api tests. The shell tests seemed like the
best way to test behavior when debug info is missing because those tests
make it easy to disable building with debug info completely.

rdar://163230807

delcypher committed:
[NFC][BoundsSafety][LLDB] Refactor the soft trap runtime tests

This refactors the soft trap runtime test so the soft trap runtime
implementation exists in its own file. This has several advantages:

* It let's the runtime be built without debug info which will be the
  common case uses hit
* It prevents the risk of infinite recursion because it isn't safe
  to build the soft trap runtime with -fbounds-safety soft trap mode
  enabled.

delcypher committed:
[BoundsSafety][LLDB] Implement instrumentation plugin for -fbounds-safety soft traps

This patch implements an instrumentation plugin for the
`-fbounds-safety` soft trap mode first implemented in
https://github.com/swiftlang/llvm-project/pull/11645 (rdar://158088757).

The current implementation of -fbounds-safety traps works by emitting
calls to runtime functions intended to log the occurrence of a soft trap.
While the user could just set a breakpoint of these functions the
instrumentation plugin sets it automatically and provides several
additional features:

When debug info is available:

* It adjusts the stop reason to be the reason for trapping. This is
  extracted from the artificial frame in the debug info (similar to
  -fbounds-safety hard traps).
* It adjusts the selected frame to be the frame where the soft trap
  occurred.

When debug info is not available:

* For the `call-with-str` soft trap mode the soft trap reason is
  read from the first argument register.
* For the `call-minimal` soft trap mode the stop reason is adjusted
  to note its a bounds check failure but does not give further
  information because none is available.
* In this situation the selected frame is not adjusted because in
  this mode the user will be looking at assembly and adjusting the
  frame makes things confusing.

This patch includes shell and api tests. The shell tests seemed like the
best way to test behavior when debug info is missing because those tests
make it easy to disable building with debug info completely.

rdar://163230807

delcypher commented:
@swift-ci test

delcypher commented:
@swift-ci test llvm

delcypher commented:
@swift-ci test

delcypher commented:
@swift-ci test llvm

delcypher commented:
[Windows failure ](https://ci-external.swift.org/job/apple-llvm-project-pull-request-windows/3256/console) looks completely unrelated to this PR:

```
21:30:12  C:\Program Files\CMake\bin\cmake.exe -B T:\x86_64-unknown-windows-msvc\DocC -S C:\Users\swift-ci\jenkins\workspace\apple-llvm-project-pull-request-windows\swift-docc -G Ninja -D ArgumentParser_DIR=T:/x86_64-unknown-windows-msvc/ArgumentParser/cmake/modules -D BUILD_SHARED_LIBS=YES -D CMAKE_BUILD_TYPE=Release -D CMAKE_C_COMPILER=T:/5/bin/clang-cl.exe -D CMAKE_C_COMPILER_TARGET=x86_64-unknown-windows-msvc -D CMAKE_C_FLAGS=/GS- /Gw /Gy /Oy /Oi /Zc:inline -D CMAKE_EXE_LINKER_FLAGS=/INCREMENTAL:NO /OPT:REF /OPT:ICF -D CMAKE_FIND_PACKAGE_PREFER_CONFIG=YES -D CMAKE_INSTALL_PREFIX=T:/Program Files/Swift/Toolchains/0.0.0+Asserts/usr -D CMAKE_MAKE_PROGRAM=C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe -D CMAKE_SHARED_LINKER_FLAGS=/INCREMENTAL:NO /OPT:REF /OPT:ICF -D CMAKE_STATIC_LIBRARY_PREFIX_Swift=lib -D CMAKE_Swift_COMPILER=T:/5/bin/swiftc.exe -D CMAKE_Swift_COMPILER_TARGET=x86_64-unknown-windows-msvc -D CMAKE_Swift_COMPILER_WORKS=YES -D CMAKE_Swift_FLAGS=-sdk \"T:/Program Files/Swift/Platforms/Windows.platform/Developer/SDKs/Windows.sdk\" -gnone -Xlinker /INCREMENTAL:NO -Xlinker /OPT:REF -Xlinker /OPT:ICF -D CMAKE_Swift_FLAGS_RELEASE=-O -D CMAKE_Swift_FLAGS_RELWITHDEBINFO=-O -D cmark-gfm_DIR=T:/Program Files/Swift/Toolchains/0.0.0+Asserts/usr/lib/cmake -D LMDB_DIR=T:/x86_64-unknown-windows-msvc/LMDB/cmake/modules -D SwiftASN1_DIR=T:/x86_64-unknown-windows-msvc/ASN1/cmake/modules -D SwiftCrypto_DIR=T:/x86_64-unknown-windows-msvc/Crypto/cmake/modules -D SwiftMarkdown_DIR=T:/x86_64-unknown-windows-msvc/Markdown/cmake/modules -D SymbolKit_DIR=T:/x86_64-unknown-windows-msvc/SymbolKit/cmake/modules
21:30:12  -- The C compiler identification is Clang 21.1.6 with MSVC-like command-line
21:30:12  -- The Swift compiler identification is Apple 6.3
21:30:13  -- Detecting C compiler ABI info
21:30:13  -- Detecting C compiler ABI info - done
21:30:13  -- Check for working C compiler: T:/5/bin/clang-cl.exe - skipped
21:30:13  -- Detecting C compile features
21:30:13  -- Detecting C compile features - done
21:30:13  CMake Error at Sources/CMakeLists.txt:11 (add_subdirectory):
21:30:13    The source directory
21:30:13  
21:30:13      C:/Users/swift-ci/jenkins/workspace/apple-llvm-project-pull-request-windows/swift-docc/Sources/SwiftDocCUtilities
21:30:13  
21:30:13    does not contain a CMakeLists.txt file.
21:30:13  
21:30:13  
21:30:13  -- Configuring incomplete, errors occurred!
21:30:13  Error: Error: cmake.exe exited with code 1.
```

likely caused by https://github.com/swiftlang/swift-docc/pull/1331. Looks like https://github.com/swiftlang/swift-docc/pull/1352 is out to revert this change.

Michael137 commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,336 @@
+private:
+  std::pair<std::string, std::optional<uint32_t>>
+  ComputeStopReasonAndSuggestedStackFrameWithDebugInfo(
+      lldb::StackFrameSP parent_sf) {
```
Can define this out-of-line in the source file?

delcypher replied:
I could but why? Is it to make looking at the class declaration easier?

delcypher replied:
I moved the larger methods into out-of-line declarations

delcypher commented:
@swift-ci please test windows platform

jimingham commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,336 @@
+    case ArchSpec::eCore_x86_32_i686:
+      // Technically some x86 calling conventions do use a register for
+      // passing the first argument but let's ignore that for now.
+      return {};
```
Would it be worth explaining here that you can't get the information for this architecture?  Otherwise the person receiving this report might waste time trying to figure out why this wasn't working in their particular case?

delcypher replied:
If LLDB has a clean way of communicating this to the user I'm willing to use it (e.g. some kind of diagnostic system). At the moment is seems like the only thing I can do is shoehorn the failure into the trap reason message and I think that would be confusing. E.g.:

```
stop reason = Soft Bounds check failed: Cannot determine trap reason on x86 architecture when debug info is missing
```

The `Cannot determ...` part of the message is supposed to be read as the specific reason for trapping normally. Reusing it here to explain why the trap reason couldn't be determined seems a bit confusing.

Ideally what is something like:

```
stop reason = Soft Bounds check failed
note: Cannot determine trap reason on x86 architecture when debug info is missing
```

but I don't think anything like this exists today.

jimingham replied:
You can write warning messages directly by writing to the stream returned by Debugger::GetAsyncErrorStream, or you can post a warning event using Debugger::ReportWarning.  I think posting a warning event is currently in favor.

delcypher replied:
I've added warnings where I think they are useful.

jimingham commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,336 @@
+    auto process = thread_sp->GetProcess();
+    if (!process)
+      return {};
+    switch (process->GetTarget().GetArchitecture().GetCore()) {
```
Not sure you need to do this for this commit, but it would be clearer if the architecture told you whether it uses registers for argument passing, than having to hard code it here.

delcypher replied:
I can file a github issue for this. Should this go to the upstream LLVM issue tracker or the Swift issue tracker?

jimingham replied:
llvm issue tracker, this is architecture and not swift specific.

delcypher replied:
https://github.com/llvm/llvm-project/issues/168602

jimingham commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,336 @@
+    // Examine the register for the first argument
+    auto *arg0_info = rc->GetRegisterInfo(
+        lldb::RegisterKind::eRegisterKindGeneric, LLDB_REGNUM_GENERIC_ARG1);
+    if (!arg0_info)
```
Along this process there are a bunch of places where it could fail.  It might help your sanity later on if you logged where the failure occurs.  Then when this is failing on a machine or for a process you can't get access to you have more of a chance of figuring out what went wrong.

delcypher replied:
What channel should I use? Presumably you mean logging code like this (first  thing I found when greping for `LLDB_LOG`) ?

``` 
Log *log = GetLog(LLDBLog::API);
LLDB_LOGF(log,
            "SBDebugger(%p)::CreateTarget (filename=\"%s\", triple=%s, "
            "platform_name=%s, add_dependent_modules=%u, error=%s) => "
            "SBTarget(%p)",
            static_cast<void *>(m_opaque_sp.get()), filename, target_triple,
            platform_name, add_dependent_modules, sb_error.GetCString(),
            static_cast<void *>(target_sp.get()));
```

jimingham replied:
TIL, there isn't an instrumentation runtime log channel, which there really should be.  So maybe better to do this as a follow-on.

delcypher replied:
@jimingham https://github.com/llvm/llvm-project/pull/168508

delcypher replied:
I've cherry-picked the upstream change into this PR and added logging.

jimingham commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,336 @@
+  return lldb::eInstrumentationRuntimeTypeBoundsSafety;
+}
+
+const RegularExpression &
```
Julian just added a InstrumentationRuntime::MatchAllModules.  You still need to override this function but if you also override MatchAllModules to return `true`, we won't waste time pointlessly trying an empty regex against all the libraries.

delcypher replied:
The override for `MatchAllModules` is declared in the header for `InstrumentationRuntimeBoundsSafety`

delcypher commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.h:
```diff
@@ -0,0 +1,61 @@
+                                  lldb::user_id_t break_id,
+                                  lldb::user_id_t break_loc_id);
+
+  bool MatchAllModules() override { return true; }
```
@jimingham The override is here. That should be sufficient right?

jimingham replied:
Ah, missed it.  Yes, that's good.

jimingham commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,336 @@
+      /*internal=*/true,
+      /*request_hardware*/ false);
+
+  if (!breakpoint)
```
Creating breakpoints really doesn't fail, so checking the breakpoint here doesn't check much.  For instance, even if the symbols returned by getBoundsSafetySoftTrapRuntimeFuncs were not present, you would still get a valid breakpoint, just one with no locations.
I am pretty sure by the time you get here you've already checked that the symbols were present (in CheckIfRuntimeIsValid) so it would be an error here if the breakpoint has no locations.  If that's right, then that would be a better check than just for the breakpoint.
This is another point where I'd add a log message so I could see right away from a log I didn't need to look further for why nothing was working...

delcypher replied:
Ah so I don't know what the interface promises but given it doesn't disallow a nullptr being returned I thought I should check for it for robustness. Should I just assert instead?

> I am pretty sure by the time you get here you've already checked that the symbols were present (in CheckIfRuntimeIsValid) so it would be an error here if the breakpoint has no locations. If that's right, then that would be a better check than just for the breakpoint.

Make sense.

delcypher replied:
Added a check/assert.

jimingham replied:
There are a few ways to get CreateBreakpoint to return an empty BreakpointSP.  For instance if you try to make a scripted breakpoint for a scripting language we don't have loaded, or pass a classname that doesn't exist.  So formally checking the SP is good.  But I was more talking about the other direction - that the breakpoint was not null does NOT mean that it succeeded in finding any targets.  For that you have to check the number of locations.

jimingham replied:
There's no way a list of function names breakpoint could not make a breakpoint, but it's still okay to check that here.  In this case, both that and the breakpoint having no locations should be an assert, since the former should never happen, and given we've just called CheckRuntimeValid which looked for those symbols, the latter shouldn't happen either.

jimingham commented on lldb/source/Plugins/InstrumentationRuntime/CMakeLists.txt:
```diff
@@ -2,6 +2,9 @@ set_property(DIRECTORY PROPERTY LLDB_PLUGIN_KIND InstrumentationRuntime)
 
 add_subdirectory(ASan)
 add_subdirectory(ASanLibsanitizers)
+# TO_UPSTREAM(BoundsSafety) ON
```
What's this for?

delcypher replied:
This is a convention we follow for `-fbounds-safety` code. It's mostly useful for people resolving merge conflicts so they can clearly see what the code causing the conflict is for.

If you would prefer to not have it I can remove it but personally I find is useful.

jimingham replied:
In this context it just leaves me confused why BoundsSafety is special in this regard?  Why isn't this around other InstrumentationRuntimes, what does it do, etc.
But also, I just did

find . -type f -exec grep TO_UPSTREAM {} \+

at the top-level llvm-project directory and got no hits.  This certainly shouldn't be the first - unexplained - instance of it.

delcypher replied:
`-fbounds-safety`'s full implementation is currently only in the swiftlang fork of `llvm-project` which this PR is targeting. If you grep that codebase you'll find plenty of `TO_UPSTREAM(` comments.

delcypher replied:
@jimingham Considering this resolved. Please re-open if it doesn't seem to be resolved.

jimingham commented on lldb/test/API/lang/BoundsSafety/soft_trap/TestBoundsSafetyInstrumentationPlugin.py:
```diff
@@ -0,0 +1,141 @@
+                                expected_func_name:str,
+                                expected_file_name:str,
+                                expected_line_num:int):
+        process = self.dbg.GetSelectedTarget().process
```
Store the process you made in the test routine in self so you don't have to recreate it here.

delcypher replied:
Fixed. I store the target and then I get the process from the target.

jimingham commented on lldb/test/API/lang/BoundsSafety/soft_trap/TestBoundsSafetyInstrumentationPlugin.py:
```diff
@@ -0,0 +1,141 @@
+        self.createTestTarget()
+        self.runCmd("run")
+
+        process = self.dbg.GetSelectedTarget().process
```
Relying on the selected target always gives me the willies.  Somebody could come along and do something clever to create another target and then this test would fail because of the wrong process.  It probably never going to cause a problem but it's bad practice.

createTestTarget returns the target that was created, and target's only have one process.  So you could remove all room for error by capturing the target that createTestTarget made and asking IT for its process.

Also, instead of getting it again from getting the selected target, it's better to set it as a ivar of the test class here, and then read it from there in the check methods you've written.

delcypher replied:
Ah I didn't know that. This was mostly a copy-and-paste of the way the existing ASan instrumentation plugin test things. I'll clean this up.

delcypher replied:
Fixed.

jimingham commented on lldb/test/API/lang/BoundsSafety/soft_trap/TestBoundsSafetyInstrumentationPlugin.py:
```diff
@@ -0,0 +1,141 @@
+        self.createTestTarget()
+        self.runCmd("run")
+
+        process = self.dbg.GetSelectedTarget().process
```
Same thing here.

delcypher replied:
Fixed.

delcypher committed:
[LLDB] Add log channel for InstrumentationRuntime plugins (#168508)

This patch adds `LLDBLog::InstrumentationRuntime` as a log channel to
provide an appropriate channel for instrumentation runtime plugins as
previously one did not exist.

A small use of the channel is added to illustrate its use. The logging
added is not intended to be comprehensive.

This is primarily motivated by an `-fbounds-safety` instrumentation
plugin (https://github.com/swiftlang/llvm-project/pull/11835).

rdar://164920875
(cherry picked from commit 46565f32d8d70f5eb9aa3aa4fba15fbd19912ccb)

delcypher commented:
Changing target from `stable/21.x` to `swift/release/6.3`.

delcypher commented:
@swift-ci test

delcypher commented:
@jimingham @Michael137 This is ready for another round of reviews.

jimingham commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,429 @@
+      /*internal=*/true,
+      /*request_hardware*/ false);
+
+  // TODO: Once we have an instrumentation log handle we should log these
```
You've added the log handle above, so this comment is outdated, right?

delcypher replied:
Oops. Good catch. Fixed it now and added a bunch of missing logging.

jimingham commented:
There's one obsolete TODO about adding an Instrumentation Log Channel which you in fact already did.  You probably do want to log the breakpoint having no locations around where the TODO is.

Other than that, LGTM.

Michael137 commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,429 @@
+
+LLDB_PLUGIN_DEFINE(InstrumentationRuntimeBoundsSafety)
+
+#define BOUNDS_SAFETY_SOFT_TRAP_MINIMAL "__bounds_safety_soft_trap"
+#define BOUNDS_SAFETY_SOFT_TRAP_S "__bounds_safety_soft_trap_s"
```
Any reason these need to be macros?
```suggestion
static constexpr llvm::StringLiteral BOUNDS_SAFETY_SOFT_TRAP_MINIMAL = "__bounds_safety_soft_trap";
static constexpr llvm::StringLiteral BOUNDS_SAFETY_SOFT_TRAP_S = "__bounds_safety_soft_trap_s";
```

Don't mind either way

delcypher replied:
I did it to make it convenient to concatenate const strings in various places. An alternative is to make the change you suggest above but use something like `llvm::raw_string_ostream` to do the concatenation. I have a slight preference for leaving it as is but let me know if you'd prefer me to change it.

Michael137 commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,429 @@
+  return Funcs;
+}
+
+#define SOFT_TRAP_CATEGORY_PREFIX "Soft "
+#define SOFT_TRAP_FALLBACK_CATEGORY                                            \
+  SOFT_TRAP_CATEGORY_PREFIX "Bounds check failed"
```
Same question re. macros. I guess in this case it's convenient for the concatenation.

delcypher replied:
Yep. Concatenation is why I'm using macros here.

Michael137 commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,429 @@
+#define BOUNDS_SAFETY_SOFT_TRAP_S "__bounds_safety_soft_trap_s"
+
+std::vector<std::string> &getBoundsSafetySoftTrapRuntimeFuncs() {
+  static std::vector<std::string> Funcs = {BOUNDS_SAFETY_SOFT_TRAP_MINIMAL,
```
Could be:
```suggestion
  static std::array<llvm::StringRef, 2> Funcs = {BOUNDS_SAFETY_SOFT_TRAP_MINIMAL,
```
?

delcypher replied:
Unfortunately that doesn't work because we need to return a reference to a `std::vector<std::string>` and that requirement for that type comes this call:

```
  auto breakpoint = process_sp->GetTarget().CreateBreakpoint(
      /*containingModules=*/nullptr,
      /*containingSourceFiles=*/nullptr, getBoundsSafetySoftTrapRuntimeFuncs(),
      eFunctionNameTypeFull, eLanguageTypeUnknown,
      /*m_offset=*/0,
      /*skip_prologue*/ eLazyBoolNo,
      /*internal=*/true,
      /*request_hardware*/ false);
```

Michael137 commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,429 @@
+
+  static lldb::StopInfoSP
+  CreateInstrumentationBoundsSafetyStopInfo(Thread &thread) {
+    return StopInfoSP(new InstrumentationBoundsSafetyStopInfo(thread));
```
I realize we probably do this in the other plugins, but can this be a `std::make_shared`?

delcypher replied:
It doesn't work because the constructor is `private`. This code is based on an existing instrumentation plugin. Presumably the constructor is private so that the only way to make the type is through the `CreateInstrumentationBoundsSafetyStopInfo` method.

```
/Volumes/user_data/xc/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__memory/shared_ptr.h:287:37: error: calling a private constructor of class 'InstrumentationBoundsSafetyStopInfo'
        ::new ((void*)__get_elem()) _Tp(_VSTD::forward<_Args>(__args)...);
                                    ^
/Volumes/user_data/xc/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__memory/shared_ptr.h:995:55: note: in instantiation of function template specialization 'std::__shared_ptr_emplace<InstrumentationBoundsSafetyStopInfo, std::allocator<InstrumentationBoundsSafetyStopInfo>>::__shared_ptr_emplace<lldb_private::Thread &>' requested here
    ::new ((void*)_VSTD::addressof(*__guard.__get())) _ControlBlock(__a, _VSTD::forward<_Args>(__args)...);
                                                      ^
/Volumes/user_data/xc/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk/usr/include/c++/v1/__memory/shared_ptr.h:1004:19: note: in instantiation of function template specialization 'std::allocate_shared<InstrumentationBoundsSafetyStopInfo, std::allocator<InstrumentationBoundsSafetyStopInfo>, lldb_private::Thread &, void>' requested here
    return _VSTD::allocate_shared<_Tp>(allocator<_Tp>(), _VSTD::forward<_Args>(__args)...);
                  ^
/Volumes/user_data/dev/llvm/upstream_swift/stable/21.x/src/llvm-project/lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:72:14: note: in instantiation of function template specialization 'std::make_shared<InstrumentationBoundsSafetyStopInfo, lldb_private::Thread &, void>' requested here
        std::make_shared<InstrumentationBoundsSafetyStopInfo>(thread));
             ^
/Volumes/user_data/dev/llvm/upstream_swift/stable/21.x/src/llvm-project/lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:295:3: note: declared private here
  InstrumentationBoundsSafetyStopInfo(Thread &thread) : StopInfo(thread, 0) {
  ^
1 error generated.

```

I could make the constructor public but then the `CreateInstrumentationBoundsSafetyStopInfo` static method becomes a little pointless.

Michael137 commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,429 @@
+bool InstrumentationRuntimeBoundsSafety::NotifyBreakpointHit(
+    void *baton, StoppointCallbackContext *context, user_id_t break_id,
+    user_id_t break_loc_id) {
+  assert(baton && "null baton");
+  if (!baton)
+    return false; ///< false => resume execution.
```
Curious about the choice of assert+return here. I understand we don't want to crash the debugger if `baton` is invalid, but if we at all believe this can be `nullptr`, what's the value of the assert? We don't usually have this pattern in LLDB. Or maybe we do but I just haven't come across it

delcypher replied:
This copied verbatim from: https://github.com/llvm/llvm-project/blob/main/lldb/source/Plugins/InstrumentationRuntime/ASan/InstrumentationRuntimeASan.cpp#L66

I think the assert is to catch an assumption being broken in debug builds but in release we just return to avoid crashing the debugger.

Michael137 commented on lldb/source/Plugins/InstrumentationRuntime/BoundsSafety/InstrumentationRuntimeBoundsSafety.cpp:
```diff
@@ -0,0 +1,429 @@
+#define BOUNDS_SAFETY_SOFT_TRAP_MINIMAL "__bounds_safety_soft_trap"
+#define BOUNDS_SAFETY_SOFT_TRAP_S "__bounds_safety_soft_trap_s"
+
+std::vector<std::string> &getBoundsSafetySoftTrapRuntimeFuncs() {
```
Any reason to return a mutable reference?

delcypher replied:
Nope. Good suggestion. I've made it `const`.

Michael137 reviewed (approved):
Apart from some nits this LGTM also

High-level question though. Why couldn't the existing verbose-trap frame recognizer be made to understand the soft trap frame? Couldn't we have just added it to the recognized regex and then re-used the existing frame recognizer to display the stop reason etc.? Is it the no-debug-info cases that are special and need the separate plugin? Makes me wonder if we could still re-use the verbose trap recognizer for the debug-info case. But might be easier said than done/might missing something completely

delcypher commented:
@Michael137 
> High-level question though. Why couldn't the existing verbose-trap frame recognizer be made to understand the soft trap frame? Couldn't we have just added it to the recognized regex and then re-used the existing frame recognizer to display the stop reason etc.? Is it the no-debug-info cases that are special and need the separate plugin? Makes me wonder if we could still re-use the verbose trap recognizer for the debug-info case. But might be easier said than done/might missing something completely

Good question. Technically we could extend the verbose-trap frame recognizer too but something this plugin does that the verbose trap frame recognizer doesn't do is set a breakpoint. With this plugin we automatically stop on a soft trap which is the behavior I wanted.

Michael137 commented:
> @Michael137
> 
> > High-level question though. Why couldn't the existing verbose-trap frame recognizer be made to understand the soft trap frame? Couldn't we have just added it to the recognized regex and then re-used the existing frame recognizer to display the stop reason etc.? Is it the no-debug-info cases that are special and need the separate plugin? Makes me wonder if we could still re-use the verbose trap recognizer for the debug-info case. But might be easier said than done/might missing something completely
> 
> Good question. Technically we could extend the verbose-trap frame recognizer too but something this plugin does that the verbose trap frame recognizer doesn't do is set a breakpoint. With this plugin we automatically stop on a soft trap which is the behavior I wanted.

Oh i see, yea we would need to stop from the frame recognizer to trigger. Technically we could make it stop at the parent frame, and then the frame recognizer kicks in? And that could be responsible for showing the stop reason? Then the frame recognizer benefits from working in the no-debug-info case too. But i'm not 100% whether this Just Works. Might be worth a try. Or if you prefer, we can merge this and then simplify/merge with the frame recognizer as a follow-up.

delcypher commented:
> Oh i see, yea we would need to stop from the frame recognizer to trigger. Technically we could make it stop at the parent frame, and then the frame recognizer kicks in? And that could be responsible for showing the stop reason?

The logic for getting the stop reason for `-fbounds-safety` soft traps is a bit complicated here. This plugin recognizes two situations:

1. Call to one of the soft trap functions where debug info is available in the caller. In this case we use the artificial stackframe in the debug info to get the stop reason (just like the verbose trap recognizer does).

2. Call to one of the soft trap functions where debug info is **not** available in the caller. In this case we inspect register state (only for one of the soft trap functions) to try to get the stop reason. This is something that the verbose frame recognizer doesn't do currently.

Pulling all that logic into the existing verbose stackframe recognizer doesn't make sense to me because `__builtin_verbose_trap()` doesn't emit code+debug info like this. It would also really over complicate the implementation just for this use case. If we did decide to move the logic into a stackframe recognizer I'd argue it probably should be its own separate one.

Given that instrumentation plugins supports the exact thing I need (set a breakpoint and when its hit do some analysis and set the stop reason) I think it's cleaner to keep this PR as is. The common logic for getting the  trap reason from the debug info has already been refactored so that both this PR and the verbose stackframe recognizer can use it so there's no unnecessary duplication of code.

> Then the frame recognizer benefits from working in the no-debug-info case too.

The only use case I see for that is if the plugin was deliberately switched off and then a breakpoint is set manually. Currently the stop reason won't be set to the trap reason. I'm not sure if that's desirable or not. If we change the stop reason it's less clear that the user's breakpoint was hit, OTOH it's harder to see the trap reason. I doubt this is a common use case because the user would need to known that the plugin exists, and how to turn it off.

> But i'm not 100% whether this Just Works. Might be worth a try. Or if you prefer, we can merge this and then simplify/merge with the frame recognizer as a follow-up.

I would prefer to just merge. I've file a radar (rdar://165066642) about considering moving some of the logic into a stackframe recognizer.

delcypher commented:
@swift-ci test

delcypher commented:
I've squash the feedback commits into the relevant commit and I've kicked off PR testing.

Michael137 commented:
> 2. Call to one of the soft trap functions where debug info is **not** available in the caller. In this case we inspect register state (only for one of the soft trap functions) to try to get the stop reason. This is something that the verbose frame recognizer doesn't do currently.

> Pulling all that logic into the existing verbose stackframe recognizer doesn't make sense to me because `__builtin_verbose_trap()` doesn't emit code+debug info like this. It would also really over complicate the implementation just for this use case. If we did decide to move the logic into a stackframe recognizer I'd argue it probably should be its own separate one.

Ok that makes sense. I naively thought we'd be able to share the logic which recognizes the stop reason in the no-debug-info case. But sounds like for `verbose_trap` that would have to look very different to the soft trap case.

> I would prefer to just merge. I'll file a radar about considering moving some of the logic into a stackframe recognizer.

Sounds good to me!

delcypher commented:
@swift-ci test windows platform

delcypher commented:
So the [windows build](https://ci-external.swift.org/job/apple-llvm-project-pull-request-windows/3306/console) is failing because the api tests don't pass there. The assert where the stop reason is checked fails. The stop reason is supposed to be `eStopReasonInstrumentation` but its `eStopReasonNone` instead. I don't know why that would be and I don't have a Windows machine to debug this so it's probably simpler to just disable the test on Windows.

```
[2025-11-20T08:00:56.506Z] FAIL: test_call_with_str_dwarf (TestBoundsSafetyInstrumentationPlugin.BoundsSafetyTestSoftTrapPlugin)

[2025-11-20T08:00:56.506Z]    Test the plugin on code built with

[2025-11-20T08:00:56.506Z] ----------------------------------------------------------------------

[2025-11-20T08:00:56.506Z] Traceback (most recent call last):

[2025-11-20T08:00:56.506Z]   File "C:\Users\swift-ci\jenkins\workspace\apple-llvm-project-pull-request-windows\llvm-project\lldb\packages\Python\lldbsuite\test\lldbtest.py", line 1843, in test_method

[2025-11-20T08:00:56.506Z]     return attrvalue(self)

[2025-11-20T08:00:56.506Z]   File "C:\Users\swift-ci\jenkins\workspace\apple-llvm-project-pull-request-windows\llvm-project\lldb\test\API\lang\BoundsSafety\soft_trap\TestBoundsSafetyInstrumentationPlugin.py", line 128, in test_call_with_str

[2025-11-20T08:00:56.506Z]     self.check_state_soft_trap_with_str(

[2025-11-20T08:00:56.506Z]   File "C:\Users\swift-ci\jenkins\workspace\apple-llvm-project-pull-request-windows\llvm-project\lldb\test\API\lang\BoundsSafety\soft_trap\TestBoundsSafetyInstrumentationPlugin.py", line 82, in check_state_soft_trap_with_str

[2025-11-20T08:00:56.506Z]     self._check_stop_reason_impl(SOFT_TRAP_FUNC_WITH_STR,

[2025-11-20T08:00:56.506Z]   File "C:\Users\swift-ci\jenkins\workspace\apple-llvm-project-pull-request-windows\llvm-project\lldb\test\API\lang\BoundsSafety\soft_trap\TestBoundsSafetyInstrumentationPlugin.py", line 22, in _check_stop_reason_impl

[2025-11-20T08:00:56.506Z]     self.assertEqual(

[2025-11-20T08:00:56.506Z] AssertionError: 1 != 10

[2025-11-20T08:00:56.506Z] Config=x86_64-T:\5\bin\clang.exe

[2025-11-20T08:00:56.506Z] ----------------------------------------------------------------------
```

delcypher commented:
@swift-ci test

delcypher commented:
@swift-ci please test Windows platform

Michael137 commented:
@charles-zablit has been working on LLDB's Windows support. Mind giving this a try on your Windows machine once you have a free second?

CC @adrian-prantl

JDevlieghere reviewed (approved):
I discussed this with Dan offline. The plugin can go upstream, only the tests depend on having bounds safety support. You can configure LLDB to use a different compiler than the in-tree one, which provides path to landing this upstream. 

I'm approving this so we can merge this PR before the branches lock later today.  I spotted some minor issues which we can address when landing this upstream.
